pkg/datapath/loader/cache.go

```diff
@@ -72,6 +109,113 @@ func (o *objectCache) Update(nodeCfg *datapath.LocalNodeConfiguration) {
+// datapath for this hash.
+func (o *objectCache) serialize(hash string) (fq *serializer.FunctionQueue, found bool) {
+	o.Lock()
+	defer o.Unlock()
+
+	fq, compiled := o.compileQueue[hash]
+	if !compiled {
+		fq = serializer.NewFunctionQueue(1)
+		o.compileQueue[hash] = fq
+	}
+	return fq, compiled
+}
+
+func (o *objectCache) lookup(hash string) (string, bool) {
+	o.Lock()
+	defer o.Unlock()
+	path, exists := o.toPath[hash]
+	return path, exists
+}
+
+func (o *objectCache) insert(hash, objectPath string) {
+	o.Lock()
+	defer o.Unlock()
+	o.toPath[hash] = objectPath
+}
+
+// build attempts to compile and cache a datapath template object file
+// corresponding to the specified endpoint configuration.
+func (o *objectCache) build(ctx context.Context, cfg *templateCfg, hash string) error {
+	templatePath := filepath.Join(o.workingDirectory, defaults.TemplatesDir, hash)
+	headerPath := filepath.Join(templatePath, common.CHeaderFileName)
+	objectPath := filepath.Join(templatePath, endpointObj)
+
+	if err := os.MkdirAll(templatePath, defaults.StateDirRights); err != nil {
+		return err
+	}
+
+	f, err := os.Create(headerPath)
+	if err != nil {
+		return &os.PathError{
+			Op:   "failed to open template header for writing",
+			Path: headerPath,
+			Err:  err,
+		}
+	}
+
+	if err = o.Datapath.WriteEndpointConfig(f, cfg); err != nil {
+		return &os.PathError{
+			Op:   "failed to write template header",
+			Path: headerPath,
+			Err:  err,
+		}
+	}
+
+	err = compileTemplate(ctx, templatePath)
+	if err != nil {
+		return err
+	}
+
+	log.WithFields(logrus.Fields{
+		logfields.Path: objectPath,
+	}).Info("Compiled new BPF template")
+
+	o.insert(hash, objectPath)
+	return nil
+}
+
+// fetchOrCompile attempts to fetch the path to the datapath object
+// corresponding to the provided endpoint configuration, or if this
+// configuration is not yet compiled, compiles it. It will block if multiple
+// threads attempt to concurrently fetchOrCompile a template binary for the
+// same set of EndpointConfiguration.
+//
+// Returns the path to the compiled template datapath object and whether the
+// object was compiled, or an error.
+func (o *objectCache) fetchOrCompile(ctx context.Context, cfg datapath.EndpointConfiguration) (string, bool, error) {
+	hash, err := o.baseHash.sumEndpoint(o, cfg, false)
+	if err != nil {
+		return "", false, err
+	}
+
+	// Serializes attempts to compile this cfg.
+	fq, compiled := o.serialize(hash)
+	if !compiled {
+		fq.Enqueue(func() error {
+			defer fq.Stop()
+			templateCfg := wrap(cfg)
+			return o.build(ctx, templateCfg, hash)
+		}, serializer.NoRetry)
+	}
+
+	// Wait until the build completes.
+	if err = fq.Wait(ctx); err != nil {
+		return "", false, fmt.Errorf("context cancelled while waiting for template compilation: %s", err)
+	}
+
+	// Fetch the result of the compilation.
+	path, ok := o.lookup(hash)
+	if !ok {
+		return "", false, fmt.Errorf("peer compilation for this template failed")
```

Comment:
I would not be sure what this error means if I saw it

Reply:
Any thoughts on a better way to express that "Another goroutine tried to compile but failed with this error."? (In retrospect maybe these exact words would be more clear... ;) )